Update the Compatibility changes of release-8.5.4 (#22233)
pingcap/docs: releases/release-8.5.4.md

## Compatibility changes

## Highlights

> 

Reviewer: ```suggestion
```

- Newly created v8.5.3 clusters can be smoothly upgraded to v8.5.4. However, this release includes **default value changes and behavior adjustments** for several system variables and configuration parameters that require attention.
  - Most changes are **safe for general upgrades**, but **clusters with customized performance tuning, TiFlash, or TiKV compaction settings should review this section carefully**.
  - Some legacy configuration items are **deprecated and no longer recommended**, and you should **migrate to the new configuration groups**.

> ## System Variables

Reviewer: ```suggestion
### System variables
```

## Compatibility changes

Newly created v8.5.3 clusters can be smoothly upgraded to v8.5.4. However, v8.5.4 introduces several **default value changes and behavior adjustments** for system variables and configuration parameters. Before upgrading, pay attention to the following:

- Most changes are safe for routine upgrades. However, if your cluster has undergone performance tuning, such as customized TiFlash and TiKV compaction configurations, review this section carefully.
- Some legacy TiKV configuration items are deprecated in v8.5.4 and no longer recommended. As a replacement, it is recommended that you use the new TiKV configuration group provided in this section.

### System variables

| Variable | Change type | Description |
|---------|------|-------------|

> | [`tidb_mpp_store_fail_ttl`](https://docs.pingcap.com/tidb/v8.5/system-variables#tidb_mpp_store_fail_ttl) | Modified | Changes the default value from `60s` to `0s`. This means TiDB no longer needs to wait before sending queries to newly started TiFlash nodes, as delays are no longer necessary to prevent query failures. [#61826](https://github.com/pingcap/tidb/issues/61826) [@gengliqi](https://github.com/gengliqi) |

Reviewer 2: ```suggestion
| [`tidb_mpp_store_fail_ttl`](https://docs.pingcap.com/tidb/v8.5/system-variables#tidb_mpp_store_fail_ttl) | Modified | Changes the default value from `60s` to `0s`. This means TiDB no longer needs to wait before sending queries to newly started TiFlash nodes, because delays are no longer necessary to prevent query failures. [#61826](https://github.com/pingcap/tidb/issues/61826) [@gengliqi](https://github.com/gengliqi) |
```

| [`tidb_replica_read`](/system-variables.md#tidb_replica_read-new-in-v40) | Modified | Starting from v8.5.4, this variable only takes effect on read-only SQL statements. This improves data read safety and reduces overlaps with other features. [#62856](https://github.com/pingcap/tidb/issues/62856) [@you06](https://github.com/you06) |
| [`tidb_opt_enable_no_decorrelate_in_select`](/system-variables.md#tidb_opt_enable_no_decorrelate_in_select-new-in-v854) | Newly added | Controls whether to decorrelate subqueries in the `SELECT` list. The default value is `OFF`. [#51116](https://github.com/pingcap/tidb/issues/51116) [@terry1purcell](https://github.com/terry1purcell) |
| [`tidb_opt_enable_semi_join_rewrite`](/system-variables.md#tidb_opt_enable_semi_join_rewrite-new-in-v854) | Newly added | Controls whether to rewrite `EXISTS` subqueries. The default value is `OFF`. [#44850](https://github.com/pingcap/tidb/issues/44850) [@terry1purcell](https://github.com/terry1purcell) |

> | [`tidb_stats_update_during_ddl`](/system-variables.md#tidb_stats_update_during_ddl-new-in-v854) | Newly added | Controls whether to enable [`ANALYZE` Embedded in DDL Statements](/ddl_embedded_analyze.md). The default value is `OFF`. When enabled, the `ADD INDEX` DDL statement collects statistics for the new index during execution, allowing the optimizer to use the index immediately after it is added. Note that enabling this variable might increase DDL execution time when adding indexes to large tables. [#57948](https://github.com/pingcap/tidb/issues/57948) [@terry1purcell](https://github.com/terry1purcell) [@AilinKid](https://github.com/AilinKid) |

Reviewer 2: ```suggestion
| [`tidb_mpp_store_fail_ttl`](https://docs.pingcap.com/tidb/v8.5/system-variables#tidb_mpp_store_fail_ttl) | Modified | Changes the default value from `60s` to `0s`. This means TiDB no longer needs to wait before sending queries to newly started TiFlash nodes, as delays are no longer necessary to prevent query failures. [#61826](https://github.com/pingcap/tidb/issues/61826) [@gengliqi](https://github.com/gengliqi) |
| [`tidb_replica_read`](https://docs.pingcap.com/tidb/v8.5/system-variables#tidb_replica_read-new-in-v40) | Modified | Starting from v8.5.4, this variable only takes effect on read-only SQL statements. This improves data read safety and reduces overlaps with other features. [#62856](https://github.com/pingcap/tidb/issues/62856) [@you06](https://github.com/you06) |
| [`tidb_opt_enable_no_decorrelate_in_select`](https://docs.pingcap.com/tidb/v8.5/system-variables#tidb_opt_enable_no_decorrelate_in_select-new-in-v854) | Newly added | Controls whether to decorrelate subqueries in the `SELECT` list. The default value is `OFF`. [#51116](https://github.com/pingcap/tidb/issues/51116) [@terry1purcell](https://github.com/terry1purcell) |
| [`tidb_opt_enable_semi_join_rewrite`](https://docs.pingcap.com/tidb/v8.5/system-variables#tidb_opt_enable_semi_join_rewrite-new-in-v854) | Newly added | Controls whether to rewrite `EXISTS` subqueries. The default value is `OFF`. [#44850](https://github.com/pingcap/tidb/issues/44850) [@terry1purcell](https://github.com/terry1purcell) |
| [`tidb_stats_update_during_ddl`](https://docs.pingcap.com/tidb/v8.5/system-variables#tidb_stats_update_during_ddl-new-in-v854) | Newly added | Controls whether to enable [`ANALYZE` Embedded in DDL Statements](https://docs.pingcap.com/tidb/v8.5/ddl_embedded_analyze). The default value is `OFF`. When enabled, the `ADD INDEX` DDL statement collects statistics for the new index during execution, allowing the optimizer to use the index immediately after it is added. Note that enabling this variable might increase DDL execution time when adding indexes to large tables. [#57948](https://github.com/pingcap/tidb/issues/57948) [@terry1purcell](https://github.com/terry1purcell) [@AilinKid](https://github.com/AilinKid) |
```

> | [`tidb_stats_update_during_ddl`](https://docs.pingcap.com/tidb/v8.5/system-variables#tidb_stats_update_during_ddl-new-in-v854) | Newly added | Controls whether to enable [`ANALYZE` Embedded in DDL Statements](https://docs.pingcap.com/tidb/v8.5/ddl_embedded_analyze). The default value is `OFF`. When enabled, the `ADD INDEX` DDL statement collects statistics for the new index during execution, allowing the optimizer to use the index immediately after it is added. Note that enabling this variable might increase DDL execution time when adding indexes to large tables. [#57948](https://github.com/pingcap/tidb/issues/57948) [@terry1purcell](https://github.com/terry1purcell) [@AilinKid](https://github.com/AilinKid) |

Reviewer 2: ```suggestion
| [`tidb_stats_update_during_ddl`](https://docs.pingcap.com/tidb/v8.5/system-variables#tidb_stats_update_during_ddl-new-in-v854) | Newly added | Controls whether to enable [`ANALYZE` Embedded in DDL Statements](https://docs.pingcap.com/tidb/v8.5/ddl_embedded_analyze). The default value is `OFF`. When it is enabled, the `ADD INDEX` DDL statement collects statistics for the new index during execution, allowing the optimizer to use the index immediately after it is added. Note that enabling this variable might increase DDL execution time when adding indexes to large tables. [#57948](https://github.com/pingcap/tidb/issues/57948) [@terry1purcell](https://github.com/terry1purcell) [@AilinKid](https://github.com/AilinKid) |
```

### Configuration parameters

| Configuration file or component | Configuration parameter | Change type | Description |
|--------------------------------|--------------------------|------|-------------|

> | TiKV | [`rocksdb.max-manifest-file-size`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#max-manifest-file-size) | Modified | Changes the default value from `128MiB` to `256MiB` to avoid frequent manifest file compactions that could affect overall performance when a single TiKV node contains a large number of SST files. [#18889](https://github.com/tikv/tikv/issues/18889) [@glorv](https://github.com/glorv) |

Reviewer 2: ```suggestion
| TiKV | [`rocksdb.max-manifest-file-size`](/tikv-configuration-file.md#max-manifest-file-size) | Modified | Changes the default value from `128MiB` to `256MiB` to avoid frequent manifest file compactions that could affect overall performance when a single TiKV node contains a large number of SST files. [#18889](https://github.com/tikv/tikv/issues/18889) [@glorv](https://github.com/glorv) |
```

| TiKV | [`grpc-raft-conn-num`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#grpc-raft-conn-num) | Modified | Changes the default value from `1` to `MAX(1, MIN(4, CPU cores / 8))`, which enables gRPC-related thread settings to adapt to the number of CPU cores. When the number of CPU cores equals or exceeds 32, the maximum default number of connections is 4. [#18806](https://github.com/tikv/tikv/issues/18806) [@LykxSassinator](https://github.com/LykxSassinator) |
| TiKV | [`grpc-concurrency`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#grpc-concurrency) | Modified | Changes the default value from `5` to `grpc-raft-conn-num * 3 + 2`, which enables gRPC-related thread settings to adapt to the number of CPU cores. [#18806](https://github.com/tikv/tikv/issues/18806) [@LykxSassinator](https://github.com/LykxSassinator) |

> | TiKV | [`region-compact-check-interval`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-check-interval), [`region-compact-check-step`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-check-step), [`region-compact-min-tombstones`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-min-tombstones), [`region-compact-tombstones-percent`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-tombstones-percent), [`region-compact-min-redundant-rows`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-min-redundant-rows-new-in-v710), [`region-compact-redundant-rows-percent`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-redundant-rows-percent-new-in-v710) | Deprecated | These configuration items are replaced by the [`gc.auto-compaction`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file/#gcauto-compaction) configuration group, which controls automatic compaction behavior. [#18727](https://github.com/tikv/tikv/issues/18727) [@v01dstar](https://github.com/v01dstar) |

Reviewer 2: ```suggestion
| TiKV | <ul><li>[`region-compact-check-interval`](/tikv-configuration-file.md#region-compact-check-interval)</li><li>[`region-compact-check-step`](/tikv-configuration-file.md#region-compact-check-step)</li><li>[`region-compact-min-tombstones`](/tikv-configuration-file.md#region-compact-min-tombstones)</li><li>[`region-compact-tombstones-percent`](/tikv-configuration-file.md#region-compact-tombstones-percent)</li><li>[`region-compact-min-redundant-rows`](/tikv-configuration-file.md#region-compact-min-redundant-rows-new-in-v710)</li><li>[`region-compact-redundant-rows-percent`](/tikv-configuration-file.md#region-compact-redundant-rows-percent-new-in-v710)</li></ul> | Deprecated | These configuration items are replaced by the [`gc.auto-compaction`](/tikv-configuration-file.md#gcauto-compaction) configuration group, which controls automatic compaction behavior. [#18727](https://github.com/tikv/tikv/issues/18727) [@v01dstar](https://github.com/v01dstar) |
```

> | TiKV | <ul><li>[`region-compact-check-interval`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-check-interval)</li><li>[`region-compact-check-step`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-check-step)</li><li>[`region-compact-min-tombstones`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-min-tombstones)</li><li>[`region-compact-tombstones-percent`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-tombstones-percent)</li><li>[`region-compact-min-redundant-rows`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-min-redundant-rows-new-in-v710)</li><li>[`region-compact-redundant-rows-percent`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-redundant-rows-percent-new-in-v710)</li></ul> | Deprecated | These configuration items are replaced by the [`gc.auto-compaction`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#gcauto-compaction) configuration group, which controls automatic compaction behavior. [#18727](https://github.com/tikv/tikv/issues/18727) [@v01dstar](https://github.com/v01dstar) |

Reviewer 2: ```suggestion
| TiKV | <ul><li>[`region-compact-check-interval`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-check-interval)</li><li>[`region-compact-check-step`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-check-step)</li><li>[`region-compact-min-tombstones`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-min-tombstones)</li><li>[`region-compact-tombstones-percent`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-tombstones-percent)</li><li>[`region-compact-min-redundant-rows`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-min-redundant-rows-new-in-v710)</li><li>[`region-compact-redundant-rows-percent`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#region-compact-redundant-rows-percent-new-in-v710)</li></ul> | Deprecated | These configuration items are replaced by the [`gc.auto-compaction`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#gcauto-compaction) configuration group, which controls the automatic compaction behavior. [#18727](https://github.com/tikv/tikv/issues/18727) [@v01dstar](https://github.com/v01dstar) |
```

> | TiKV | [`gc.auto-compaction`](/tikv-configuration-file.md#gcauto-compaction) configuration group: <ul><li>[`gc.auto-compaction.check-interval`](/tikv-configuration-file.md#check-interval-new-in-v757-and-v854)</li><li>[`gc.auto-compaction.tombstone-num-threshold`](/tikv-configuration-file.md#tombstone-num-threshold-new-in-v757-and-v854), [`gc.auto-compaction.tombstone-percent-threshold`](/tikv-configuration-file.md#tombstone-percent-threshold-new-in-v757-and-v854)</li><li>[`gc.auto-compaction.redundant-rows-threshold`](/tikv-configuration-file.md#redundant-rows-threshold-new-in-v757-and-v854)</li><li>[`gc.auto-compaction.redundant-rows-percent-threshold`](/tikv-configuration-file.md#redundant-rows-percent-threshold-new-in-v757-and-v854)</li><li>[`gc.auto-compaction.bottommost-level-force`](/tikv-configuration-file.md#bottommost-level-force-new-in-v757-and-v854) </li></ul>| New | This configuration group controls automatic compaction behavior. [#18727](https://github.com/tikv/tikv/issues/18727) [@v01dstar](https://github.com/v01dstar) |

Reviewer 2: ```suggestion
| TiKV | [`gc.auto-compaction`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#gcauto-compaction) configuration group: <ul><li>[`gc.auto-compaction.check-interval`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#check-interval-new-in-v757-and-v854)</li><li>[`gc.auto-compaction.tombstone-num-threshold`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#tombstone-num-threshold-new-in-v757-and-v854), [`gc.auto-compaction.tombstone-percent-threshold`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#tombstone-percent-threshold-new-in-v757-and-v854)</li><li>[`gc.auto-compaction.redundant-rows-threshold`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#redundant-rows-threshold-new-in-v757-and-v854)</li><li>[`gc.auto-compaction.redundant-rows-percent-threshold`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#redundant-rows-percent-threshold-new-in-v757-and-v854)</li><li>[`gc.auto-compaction.bottommost-level-force`](https://docs.pingcap.com/tidb/v8.5/tikv-configuration-file#bottommost-level-force-new-in-v757-and-v854) </li></ul>| New | This configuration group controls automatic compaction behavior. [#18727](https://github.com/tikv/tikv/issues/18727) [@v01dstar](https://github.com/v01dstar) |
```
